File: src/cyberrunner_vision/cyberrunner_vision/camera_node.py
"""
camera_node  (Linux-compatible)
--------------------------------
Parameters
  camera_index  int     default 0
  device_path   str     default ""   e.g. "/dev/video0"  (overrides camera_index)
  fx            float   default 850.0
  fy            float   default 750.0
  width         int     default 1920
  height        int     default 1200
  fps           int     default 30
"""
import sys
import rclpy
from rclpy.node import Node
from sensor_msgs.msg import Image
from cv_bridge import CvBridge
import cv2
import numpy as np
from .ocam_params import OCAM, OUT_W, OUT_H


# ── OCamCalib rectification ───────────────────────────────────────────────────
def _world2cam(dirs, ocam):
    X, Y, Z = dirs[..., 0], dirs[..., 1], dirs[..., 2]
    rho = np.sqrt(X * X + Y * Y) + 1e-12
    t   = np.arctan2(-Z, rho)
    r   = np.zeros_like(t)
    tt  = np.ones_like(t)
    for a in ocam["invpol"]:
        r += a * tt
        tt *= t
    x = X / rho;  y = Y / rho
    u = ocam["cx"] + (x * r) * ocam["c"] + (y * r) * ocam["d"]
    v = ocam["cy"] + (x * r) * ocam["e"] + (y * r)
    return u.astype(np.float32), v.astype(np.float32)


def build_rectify_map(ocam, out_w, out_h, fx, fy):
    uu, vv = np.meshgrid(np.arange(out_w, dtype=np.float64),
                         np.arange(out_h, dtype=np.float64))
    x = (uu - out_w / 2) / fx
    y = (vv - out_h / 2) / fy
    z = np.ones_like(x)
    n = np.sqrt(x * x + y * y + z * z)
    dirs = np.stack([x / n, y / n, z / n], axis=-1)
    return _world2cam(dirs, ocam)


# ── Open camera (Linux-safe) ──────────────────────────────────────────────────
def open_camera(index: int, device_path: str, width: int, height: int, fps: int):
    """
    Try multiple backends in order until one works.
    On Linux: V4L2 first, then ANY.
    device_path overrides index when non-empty.
    """
    source = device_path if device_path else index

    # Build explicit GStreamer pipeline (most reliable for this camera)
    gst = (
        f"v4l2src device={source if isinstance(source, str) else '/dev/video'+str(source)} "
        f"! image/jpeg,width={width},height={height},framerate={fps}/1 "
        f"! jpegdec ! videoconvert ! appsink"
    )
    cap = cv2.VideoCapture(gst, cv2.CAP_GSTREAMER)
    if cap.isOpened():
        for _ in range(3):
            cap.grab()
        ok, _ = cap.read()
        if ok:
            return cap, "gstreamer-pipeline"
        cap.release()

    # Fallback: V4L2 with MJPG
    cap = cv2.VideoCapture(source, cv2.CAP_V4L2)
    if cap.isOpened():
        cap.set(cv2.CAP_PROP_FOURCC, cv2.VideoWriter_fourcc(*'MJPG'))
        if width > 0:  cap.set(cv2.CAP_PROP_FRAME_WIDTH,  width)
        if height > 0: cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
        if fps > 0:    cap.set(cv2.CAP_PROP_FPS,          fps)
        for _ in range(3):
            cap.grab()
        ok, _ = cap.read()
        if ok:
            return cap, "v4l2"
        cap.release()

    return None, None


# ── Node ─────────────────────────────────────────────────────────────────────
class CameraNode(Node):
    def __init__(self):
        super().__init__("camera_node")
        self.declare_parameter("camera_index",  0)
        self.declare_parameter("device_path",   "")     # e.g. "/dev/video0"
        self.declare_parameter("fx",      850.0)
        self.declare_parameter("fy",      750.0)
        self.declare_parameter("width",   OCAM["width"])
        self.declare_parameter("height",  OCAM["height"])
        self.declare_parameter("fps",     30)
        self.declare_parameter("show_preview", True)    # set False on headless machines

        idx    = self.get_parameter("camera_index").value
        dpath  = self.get_parameter("device_path").value
        fx     = self.get_parameter("fx").value
        fy     = self.get_parameter("fy").value
        width  = self.get_parameter("width").value
        height = self.get_parameter("height").value
        fps    = self.get_parameter("fps").value

        self.map_x, self.map_y = build_rectify_map(OCAM, OUT_W, OUT_H, fx, fy)
        self.bridge = CvBridge()

        self.cap, backend = open_camera(idx, dpath, width, height, fps)

        if self.cap is None:
            self.get_logger().error(
                f"Cannot open camera (index={idx}, device_path='{dpath}'). "
                f"Run:  ls /dev/video*   to find the correct device.")
            raise RuntimeError("Camera open failed")

        actual_w = int(self.cap.get(cv2.CAP_PROP_FRAME_WIDTH))
        actual_h = int(self.cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
        src_str = dpath if dpath else str(idx)
        self.get_logger().info(
            "Camera opened  source=" + src_str +
            "  backend=" + str(backend) +
            "  resolution=" + str(actual_w) + "x" + str(actual_h) +
            "  fx=" + str(fx) + "  fy=" + str(fy))

        self.show_preview = bool(self.get_parameter("show_preview").value)
        self.pub   = self.create_publisher(Image, "/camera/rectified", 2)
        self.timer = self.create_timer(1.0 / fps, self._tick)

    def _tick(self):
        ok, frame = self.cap.read()
        if not ok:
            self.get_logger().warn("Camera read failed", throttle_duration_sec=2.0)
            return
        rect = cv2.remap(frame, self.map_x, self.map_y,
                         cv2.INTER_LINEAR,
                         borderMode=cv2.BORDER_CONSTANT)
        msg = self.bridge.cv2_to_imgmsg(rect, encoding="bgr8")
        msg.header.stamp    = self.get_clock().now().to_msg()
        msg.header.frame_id = "camera"
        self.pub.publish(msg)

        if self.show_preview:
            cv2.imshow("RECTIFIED", rect)
            cv2.waitKey(1)

    def destroy_node(self):
        if self.cap:
            self.cap.release()
        cv2.destroyAllWindows()
        super().destroy_node()


def main(args=None):
    rclpy.init(args=args)
    node = CameraNode()
    try:
        rclpy.spin(node)
    finally:
        node.destroy_node()
        rclpy.shutdown()

File: src/cyberrunner_cpp/src/camera_node.cpp
/**
 * camera_node  (C++ port of camera_node.py)
 * ------------------------------------------
 * Captures frames from a fisheye camera, rectifies them with the OCamCalib
 * inverse-polynomial model, and publishes to /camera/rectified.
 *
 * Parameters
 *   camera_index  int     default 0
 *   device_path   string  default ""  e.g. "/dev/video0"  (overrides camera_index)
 *   fx            double  default 850.0
 *   fy            double  default 750.0
 *   width         int     default 1920
 *   height        int     default 1200
 *   fps           int     default 114
 *   show_preview  bool    default true
 */
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>
#include "cyberrunner/ocam_model.hpp"

class CameraNode : public rclcpp::Node
{
public:
    CameraNode() : Node("camera_node")
    {
        this->declare_parameter("camera_index",  0);
        this->declare_parameter("device_path",   std::string(""));
        this->declare_parameter("fx",            OCAM_DEFAULT_FX);
        this->declare_parameter("fy",            OCAM_DEFAULT_FY);
        this->declare_parameter("width",         ocam_.width);
        this->declare_parameter("height",        ocam_.height);
        this->declare_parameter("fps",           114);
        this->declare_parameter("show_preview",  true);

        int         idx   = this->get_parameter("camera_index").as_int();
        std::string dpath = this->get_parameter("device_path").as_string();
        double      fx    = this->get_parameter("fx").as_double();
        double      fy    = this->get_parameter("fy").as_double();
        int         width = this->get_parameter("width").as_int();
        int         height= this->get_parameter("height").as_int();
        int         fps   = std::max(0, static_cast<int>(this->get_parameter("fps").as_int()));
        show_preview_     = this->get_parameter("show_preview").as_bool();

        // Pre-compute rectification maps
        buildOcamRectifyMap(ocam_, OCAM_OUT_W, OCAM_OUT_H, fx, fy, map_x_, map_y_);

        // Open the camera
        openCamera(idx, dpath, width, height, fps);

        if (!cap_.isOpened()) {
            RCLCPP_ERROR(this->get_logger(),
                "Cannot open camera (index=%d, device_path='%s'). Check: ls /dev/video*",
                idx, dpath.c_str());
            throw std::runtime_error("Camera open failed");
        }

        double actual_fps = cap_.get(cv::CAP_PROP_FPS);
        double timer_fps  = (std::isfinite(actual_fps) && actual_fps > 1.0)
                                ? actual_fps
                                : static_cast<double>(fps > 0 ? fps : 30);

        RCLCPP_INFO(this->get_logger(),
            "Camera opened  actual_fps=%.1f  timer_fps=%.1f  fx=%.1f  fy=%.1f",
            actual_fps, timer_fps, fx, fy);

        pub_   = this->create_publisher<sensor_msgs::msg::Image>("/camera/rectified", 2);
        timer_ = this->create_wall_timer(
            std::chrono::duration<double>(1.0 / timer_fps),
            std::bind(&CameraNode::tick, this));
    }

    ~CameraNode() override
    {
        if (cap_.isOpened()) cap_.release();
        cv::destroyAllWindows();
    }

private:
    void openCamera(int idx, const std::string& dpath, int width, int height, int fps)
    {
        std::string dev = dpath.empty()
                            ? "/dev/video" + std::to_string(idx)
                            : dpath;

        // Try GStreamer pipelines (with fps cap first, then without)
        std::vector<std::string> caps_list;
        if (fps > 0) {
            caps_list.push_back(
                "image/jpeg,width=" + std::to_string(width) +
                ",height=" + std::to_string(height) +
                ",framerate=" + std::to_string(fps) + "/1");
        }
        caps_list.push_back(
            "image/jpeg,width=" + std::to_string(width) +
            ",height=" + std::to_string(height));

        for (const auto& caps : caps_list) {
            std::string gst =
                "v4l2src device=" + dev +
                " ! " + caps +
                " ! jpegdec ! videoconvert"
                " ! appsink drop=true max-buffers=1 sync=false";
            cap_.open(gst, cv::CAP_GSTREAMER);
            if (cap_.isOpened()) {
                for (int i = 0; i < 3; ++i) cap_.grab();
                cv::Mat tmp;
                if (cap_.read(tmp) && !tmp.empty()) {
                    RCLCPP_INFO(this->get_logger(), "Opened via GStreamer");
                    return;
                }
                cap_.release();
            }
        }

        // Fallback: V4L2 with MJPG
        if (!dpath.empty()) {
            cap_.open(dpath, cv::CAP_V4L2);
        } else {
            cap_.open(idx, cv::CAP_V4L2);
        }
        if (cap_.isOpened()) {
            cap_.set(cv::CAP_PROP_FOURCC, cv::VideoWriter::fourcc('M', 'J', 'P', 'G'));
            if (width > 0)  cap_.set(cv::CAP_PROP_FRAME_WIDTH,  width);
            if (height > 0) cap_.set(cv::CAP_PROP_FRAME_HEIGHT, height);
            if (fps > 0)    cap_.set(cv::CAP_PROP_FPS,          fps);
            for (int i = 0; i < 3; ++i) cap_.grab();
            cv::Mat tmp;
            if (cap_.read(tmp) && !tmp.empty()) {
                RCLCPP_INFO(this->get_logger(), "Opened via V4L2/MJPG");
                return;
            }
            cap_.release();
        }
    }

    void tick()
    {
        cv::Mat frame;
        if (!cap_.read(frame) || frame.empty()) {
            RCLCPP_WARN_THROTTLE(this->get_logger(), *this->get_clock(), 2000,
                "Camera read failed");
            return;
        }

        cv::Mat rect;
        cv::remap(frame, rect, map_x_, map_y_, cv::INTER_LINEAR, cv::BORDER_CONSTANT);

        auto msg = cv_bridge::CvImage(
            std_msgs::msg::Header{}, "bgr8", rect).toImageMsg();
        msg->header.stamp    = this->now();
        msg->header.frame_id = "camera";
        pub_->publish(*msg);

        if (show_preview_) {
            cv::imshow("RECTIFIED", rect);
            cv::waitKey(1);
        }
    }

    OcamModel        ocam_;
    cv::VideoCapture cap_;
    cv::Mat          map_x_, map_y_;
    bool             show_preview_{true};

    rclcpp::Publisher<sensor_msgs::msg::Image>::SharedPtr pub_;
    rclcpp::TimerBase::SharedPtr                         timer_;
};

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CameraNode>();
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}
